tensorflow_datasets/scripts/cleanup/url_filename_recorder.py

```diff
@@ -0,0 +1,77 @@
+# coding=utf-8
+# Copyright 2020 The TensorFlow Datasets Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""Update checksums to contain filename."""
+
+import os
+import pathlib
+import requests
+from absl import app
+from typing import Dict, List
+from concurrent import futures
+
+import tensorflow_datasets as tfds
+from tensorflow_datasets.core.download import checksums
+
+
+TFDS_PATH = tfds.core.utils.tfds_write_path()
+
+
+def _collect_path_to_update() -> List[tfds.core.utils.ReadWritePath]:
+  """Collect checksums paths to update."""
+  url_checksums_paths = list(checksums._checksum_paths().values())
+
+  builder_checksums_path = []
+  for name in tfds.list_builders():
+    builder_cls = tfds.builder_cls(name)
+    if builder_cls.url_infos != None:
+      builder_checksums_path.append(builder_cls._checksums_path)
+  return url_checksums_paths + builder_checksums_path
+
+def _request_filename(url) -> (str, str):
+  """Get filename of dataset at `url`."""
+  filename  = ''
+  try:
+    response = requests.get(url, timeout=10)
+    filename = tfds.core.download.downloader._get_filename(response)
+    print(f'Success for {url}')
+  except requests.exceptions.HTTPError as http_err:
+    print(f'HTTP Error {http_err} for {url}.')
+  return (url, filename)
+
+def _update_url_infos(url_infos, filenames) -> Dict[str, checksums.UrlInfo]:
+  """Get and update dataset filname in UrlInfo."""
+  for url, filename in filenames:
+    url_infos[url].filename = filename
+  return url_infos
+
+
+def main(_):
+  with futures.ThreadPoolExecutor(max_workers=100) as executor:
```

Comment:
How about my suggestion from the previous message ? I updated the message aftwerwrd so you might not have been notified:

```py
path_to_url_infos = _collect_path_to_url_infos() -> Dict[path, Dict[url, UrlInfo]]

# Flatten all urls to remove duplicates across datasets
all_url_infos = {}
for url_infos in path_to_url_infos.values()
  all_url_infos.update(url_infos)

with futures.ThreadPoolExecutor(max_workers=100) as executor:
  # Process all urls in parallel
  url_to_filename = {
      url: executor.submit(_get_filename, url, url_info)
      for url, url_info in all_url_infos.items()
  }

  # Then update path_to_url_infos with filenames
  for path, url_infos in in path_to_url_infos.items():
    ...
```

Reply:
Oh! We want to prevent duplicate urls. Sounds great. Working on it, will update as soon as I'm done.

Thanks a lot.

Reply:
I am not able to understand one thing. According to the method that you have suggested, I think we need to wait for the async call to to finish, and when we get all the filenames then we can start updating `url_infos` with filenames, this makes the threading useless. So, if we need to accomplish it, I think we need one more dictionary to map `url->List[paths]`. 
I'm working on it. Please do let me know.

Thanks.